nexus/src/db/datastore.rs

```diff
@@ -1602,19 +1608,24 @@ impl DataStore {
                 insert_new_query.insert_and_get_results(conn).map_err(|e| {
                     match e {
                         SyncInsertError::CollectionNotFound => {
-                            diesel::result::Error::RollbackTransaction
+                            TxnError::CustomError(
+                                FirewallUpdateError::CollectionNotFound,
+                            )
                         }
-                        SyncInsertError::DatabaseError(e) => e,
+                        SyncInsertError::DatabaseError(e) => e.into(),
                     }
                 })
             })
             .await
-            .map_err(|e| {
-                public_error_from_diesel_pool(
+            .map_err(|e| match e {
+                TxnError::CustomError(
+                    FirewallUpdateError::CollectionNotFound,
```

Comment:
Are you only using `FirewallUpdateError::CollectionNotFound` to match on here? Could this be done with `SyncInsertError::CollectionNotFound` instead?

Reply:
I need the error type returned from `.transaction` to impl both:

- `From<PoolError>`
- `From<DieselError>`

The `TxnError` type (using the generic `TransactionError` enum) does that, but `SyncInsertError` does not.

I don't think it makes sense to go back to a `SyncInsertError` type here - I get that there's some redundancy between the `FirewallUpdateError` enum and the `SyncInsertError::CollectionNotFound` enum variant, but since it's just an empty marker, I figured it wasn't worth refactoring out.

Reply:
I was picturing this — so, still using `TxnError`, but eliminating `FirewallUpdateError` — but by trying it I was able to see why it doesn't work: the match gets mad at you for not handling `TxnError::CustomError(SyncInsertError::DatabaseError)`. So understand better why `FirewallUpdateError` exists — it's to express that you're never going to see `TxnError::CustomError(SyncInsertError::DatabaseError)`.

```diff
diff --git a/nexus/src/db/datastore.rs b/nexus/src/db/datastore.rs
index f7b9df2..334a96f 100644
--- a/nexus/src/db/datastore.rs
+++ b/nexus/src/db/datastore.rs
@@ -1588,11 +1588,7 @@ impl DataStore {
             diesel::insert_into(dsl::vpc_firewall_rule).values(rules),
         );
 
-        #[derive(Debug)]
-        enum FirewallUpdateError {
-            CollectionNotFound,
-        }
-        type TxnError = TransactionError<FirewallUpdateError>;
+        type TxnError = TransactionError<SyncInsertError>;
 
         // TODO-scalability: Ideally this would be a CTE so we don't need to
         // hold a transaction open across multiple roundtrips from the database,
@@ -1609,7 +1605,7 @@ impl DataStore {
                     match e {
                         SyncInsertError::CollectionNotFound => {
                             TxnError::CustomError(
-                                FirewallUpdateError::CollectionNotFound,
+                                SyncInsertError::CollectionNotFound,
                             )
                         }
                         SyncInsertError::DatabaseError(e) => e.into(),
@@ -1618,9 +1614,9 @@ impl DataStore {
             })
             .await
             .map_err(|e| match e {
-                TxnError::CustomError(
-                    FirewallUpdateError::CollectionNotFound,
-                ) => Error::not_found_by_id(ResourceType::Vpc, vpc_id),
+                TxnError::CustomError(SyncInsertError::CollectionNotFound) => {
+                    Error::not_found_by_id(ResourceType::Vpc, vpc_id)
+                }
                 TxnError::Pool(e) => public_error_from_diesel_pool(
                     e,
                     ResourceType::VpcFirewallRule,
```